middlewares/osio/auth.go

```diff
@@ -24,67 +46,138 @@ type cacheData struct {
+		if err != nil {
+			log.Errorf("Failed to locate tenant, %v", err)
+			return cacheData{}, err
+		}
+		loc := ns.ClusterURL
+		osoProxySAToken, err := srvAccTokenLocator()
+		if err != nil {
+			log.Errorf("Failed to locate service account token, %v", err)
+			return cacheData{}, err
+		}
+		clusterToken, err := tokenLocator.GetTokenWithSAToken(osoProxySAToken, loc)
+		if err != nil {
+			log.Errorf("Failed to locate cluster token, %v", err)
+			return cacheData{}, err
+		}
+		secretName, err := secretLocator.GetName(ns.ClusterURL, clusterToken, ns.Name, ns.Type)
 		if err != nil {
+			log.Errorf("Failed to locate secret name, %v", err)
 			return cacheData{}, err
 		}
-		osoToken, err := tokenLocator(osioToken, loc)
+		osoToken, err := secretLocator.GetSecret(ns.ClusterURL, clusterToken, ns.Name, secretName)
 		if err != nil {
+			log.Errorf("Failed to get secret, %v", err)
 			return cacheData{}, err
 		}
 		return cacheData{Location: loc, Token: osoToken}, nil
 	}
 }
 
-func (a *OSIOAuth) resolve(osioToken string) (cacheData, error) {
-	key := cacheKey(osioToken)
-	val, err := a.cache.Get(key, cacheResolver(a.RequestTenantLocation, a.RequestTenantToken, osioToken)).Get()
+func cacheResolverByToken(tenantLocator TenantLocator, tokenLocator TenantTokenLocator, token string) Resolver {
+	return func() (interface{}, error) {
+		ns, err := tenantLocator.GetTenant(token)
+		if err != nil {
+			log.Errorf("Failed to locate tenant, %v", err)
+			return cacheData{}, err
+		}
+		loc := ns.ClusterURL
+		osoToken, err := tokenLocator.GetTokenWithUserToken(token, loc)
+		if err != nil {
+			log.Errorf("Failed to locate token, %v", err)
+			return cacheData{}, err
+		}
+		return cacheData{Location: loc, Token: osoToken}, nil
+	}
+}
+
+func (a *OSIOAuth) resolveByToken(token string) (cacheData, error) {
+	key := cacheKey(token)
+	val, err := a.cache.Get(key, cacheResolverByToken(a.RequestTenantLocation, a.RequestTenantToken, token)).Get()
+
+	if data, ok := val.(cacheData); ok {
+		return data, err
+	}
+	return cacheData{}, err
+}
+
+func (a *OSIOAuth) resolveByID(userID, token string) (cacheData, error) {
+	key := cacheKey(userID)
+	val, err := a.cache.Get(key, cacheResolverByID(a.RequestTenantLocation, a.RequestTenantToken, a.RequestSrvAccToken, a.RequestSecretLocation, token, userID)).Get()
 
 	if data, ok := val.(cacheData); ok {
 		return data, err
 	}
 	return cacheData{}, err
 }
 
-//
 func (a *OSIOAuth) ServeHTTP(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+
 	if a.RequestTenantLocation != nil {
 
 		if r.Method != "OPTIONS" {
-			osioToken, err := getToken(r)
+			token, err := getToken(r)
 			if err != nil {
+				log.Errorf("Token not found, %v", err)
 				rw.WriteHeader(http.StatusUnauthorized)
 				return
 			}
 
-			cached, err := a.resolve(osioToken)
+			var cached cacheData
+			service := getService(r)
+			if service == che {
+				userID := r.Header.Get(impersonate)
+				cached, err = a.resolveByID(userID, token)
```

Comment:
I think you need to combine the token and the ID as the cache key.

In the current form, it seems like if a request to /api is done against id X, the token is validated by the call to Tenant Service (rh-che sa). But after that looked up in Cache based on just X.

Which would mean that as long as a valid user has been using Che at some point, _anyone_ with _any_ authorization header can later reuse that users auth by simply passing in a Authorization header (could contain anything) and a "impersonate=X" header.

Reply:
yup, we should use key=token+userID here :+1: